plugins/thunderbolt/fu-thunderbolt-image.c

```diff
@@ -439,20 +493,36 @@ get_host_locations (guint16 id)
 }
 
 static const FuThunderboltFwLocation *
-get_device_locations (guint16 id)
+get_device_locations (guint16 id, const FuThunderboltFwObject *fw, GError **error)
 {
-	static const FuThunderboltFwLocation locations[] = {
+	static const FuThunderboltFwLocation AR[] = {
 		{ .offset = 0x124, .len = 1, .section = ARC_PARAMS_SECTION, .description = "X of N" },
 		{ 0 }
 	};
 
+	static FuThunderboltFwLocation TR[] = {
+		{ .offset = 0x45,  .len = 1, .description = "Flash Size", .mask = 0x07 },
+		/* offset and len are read from DROM */
+		{ .section = DROM_SECTION, .description = "Multi Controller" },
+		{ 0 }
+	};
+
 	switch (id) {
 	case 0x1578:
 	case 0x1576:
 	case 0x15D3:
 	case 0x15DA:
 	case 0x15C0:
-		return locations;
+		return AR;
+	case 0x15E7:
+	case 0x15EA:
+	case 0x15EF:
+		/* if the controller has multi controller entry need to
+		 * compare it against the image. */
+		read_drom_entry_location (fw, DROM_ENTRY_MC, &TR[1], error);
+		if (*error == NULL)
+			return TR;
```

Comment:
this looks fishy; error may be NULL. Do you maybe want to use `error_local` here?

Reply:
Right, I'll change that to use error_local.

Reply:
Ahh, so maybe I should have been more verbose, sorry. The GLib convention for returning errors is to have a return code and a `GError**`. If the return code is FALSE then you have to set the error. If the return code is TRUE then you cannot set the error. It's also very unusual to test the error to see if the function failed, it's much more usual to test the actual return code, and then "propagate" the error if so. Would you like me to fix up the branch for you?

Reply:
Thanks for the explanation. I can fix the branch myself, though :) 

Something like:
```
if (read_drom_entry_location (fw, DROM_ENTRY_MC, &TR[1], error))
        return TR;
```
and then make `read_drom_entry_location()` return FALSE in case of error (and fill the error)?

Reply:
If i understand correctly, read_drom_entry_location is optional, so if it fails it's no big deal. At the moment you're setting the passed in GError if the read_drom_entry_location fails -- so I think you probably want to do something like:

    g_autoptr(GError) error_local = NULL;
    if (read_drom_entry_location (fw, DROM_ENTRY_MC, &TR[1], &error_local))
        return TR;
    g_debug ("failed to parse DROM something: %s", error_local->message);